### PR TITLE
feat: RocksDB Backend Support

### LINKED ISSUE #20: RocksDB Adapter
I'm looking into making a **RocksDB** adapter for this library (I can't stand **Sled** and I don't trust it). 

If I end up getting around to doing it, would this be something you would be interested in adding to the main repo, or would you rather me just fork the project and have it be it's own thing?

### PR DESCRIPTION
**Pull request by HyperAlch**
This adds support for **RocksDB** as an alternative backend for non-wasm based environments. The default backend, **Sled**, is still used unless `rocksdb` is enabled as a feature.

The test for the `clear` function was also altered to cover the way the **RocksDB** adapter implements it.

Closes #20

### CODE REVIEW

**Comment by HyperAlch** [comment]
I just realized that I forgot to create a separate branch for this. Let me know if you would like me to resubmit this using a new branch.

File: Cargo.toml [hunk 1]
```diff
@@ -12,11 +12,13 @@ repository = "https://github.com/johanhelsing/bevy_pkv"
 [dependencies]
 thiserror = "1.0"
 serde = "1.0"
+rocksdb = "0.20.1"
 bevy_ecs = "0.10" # we need for deriving Resource in PkvStore
 
 [features]
 default = ["bevy"]
 bevy = []
+rocksdb = []
```
Comment (johanhelsing) [inline Cargo.toml:21]:
```suggestion
rocksdb = ["dep:rocksdb"]
sled = ["dep:sled"]
```

File: Cargo.toml [hunk 2]
```diff
@@ -12,11 +12,13 @@ repository = "https://github.com/johanhelsing/bevy_pkv"
 [dependencies]
 thiserror = "1.0"
 serde = "1.0"
+rocksdb = "0.20.1"
```
Comment (johanhelsing) [inline Cargo.toml:15]:
```suggestion
rocksdb = { version = "0.20", optional = true }
sled = { version = "0.34", optional = true }
```
Make the dependencies optional. Also, is there a reason you have to depend on 0.20.1, and not 0.20?

File: src/rocksdb_store.rs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    /// Clear all keys and their values
+    /// The RocksDB adapter uses an iterator to achieve this, unlike sled
+    fn clear(&mut self) -> Result<(), Self::SetError> {
+        let kv_iter = self.db.iterator(rocksdb::IteratorMode::Start);
+
+        for kv in kv_iter {
+            let result = kv?;
+            let key = result.0;
```
Comment (johanhelsing) [inline src/rocksdb_store.rs:103]:
```suggestion
            let (key, _) = kv?;
```

File: src/rocksdb_store.rs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            let key = std::str::from_utf8(&key)?;
```
Comment (johanhelsing) [inline src/rocksdb_store.rs:104]: Is this necessary? Isn't the key just going to get converted back into a u8 slice?

File: Cargo.toml [hunk 5]
```diff
@@ -12,11 +12,13 @@ repository = "https://github.com/johanhelsing/bevy_pkv"
...
 bevy_ecs = "0.10" # we need for deriving Resource in PkvStore
 
 [features]
 default = ["bevy"]
```
Comment (johanhelsing) [inline Cargo.toml:19]: Hmm... Is there a way we could make default features enable sled on native, but not also break wasm builds?

**Review by johanhelsing — CHANGES_REQUESTED** [review]
Cool, implementation looks solid. Also nice with some more coverage in the test.

Currently, the rocksdb dependency is always included, and will break wasm builds, so this has to be fixed somehow.

File: src/rocksdb_store.rs (hunk 4 above)
Comment (HyperAlch) [reply src/rocksdb_store.rs:104]: Line 104 is not necessary, your right. I don't know where my brain was with that one.

File: Cargo.toml (hunk 2 above)
Comment (HyperAlch) [reply Cargo.toml:15]: There is no real reason, I'm not sure why I did that. Thanks for fixing it.

File: Cargo.toml (hunk 5 above)
Comment (HyperAlch) [reply Cargo.toml:20]: This concern was fixed with `rocksdb = { version = "0.20", optional = true }` above correct?

File: src/rocksdb_store.rs (hunk 4 above)
Comment (HyperAlch) [reply src/rocksdb_store.rs:103]: Also, if we remove `let key = std::str::from_utf8(&key)?;` we can also remove:

```rust
#[error("Invalid RocksDB key name error")]
RocksdbKeyName(#[from] Utf8Error),
```

File: Cargo.toml (hunk 5 above)
Comment (johanhelsing) [reply Cargo.toml:20]: well, it does, but it also means sled will still be a dependency when the `rocksdb` feature is enabled... `sled` should somehow be an optional dependency.

Comment (johanhelsing) [reply Cargo.toml:20]: We could make `local_storage` a feature as well, but that would require people to manually select the backend they would like, which would be more annoying to use.

Comment (HyperAlch) [reply Cargo.toml:20]: https://stackoverflow.com/questions/67373380/can-i-set-cargo-projects-default-features-depending-on-the-platform

Is this a solution?

Comment (HyperAlch) [reply Cargo.toml:20]: The OP in that Stack Overflow post didn't seem to think it was a good solution so maybe not...

Comment (johanhelsing) [reply Cargo.toml:20]: yeah, that's actually what we're currently doing, but now we're running into the issues mentioned in the last comment: https://stackoverflow.com/questions/67373380/can-i-set-cargo-projects-default-features-depending-on-the-platform#comment124334292_70316751

Comment (johanhelsing) [reply Cargo.toml:20]: I think most crates just make native work, and then tell their wasm users to deal with it by removing default-features and adding a default feature.

At least that what `surf` does: https://github.com/http-rs/surf/blob/main/Cargo.toml

Could maybe check some more cross-platform crates for inspiration.

Comment (HyperAlch) [reply Cargo.toml:20]: That could work. Or worse comes to worse maybe I make the RocksDB version a fork of your project and we just link each other on the README files? Then I would just re-implement any changes and updates you push from sled to rocksdb when needed?

Once again that would be worse case scenario. It's up to you how you want to do it. I don't want to come in and change how you do things after you have put the work in time into this library.

Comment (johanhelsing) [reply Cargo.toml:20]: winit solves this by having unsupported features just be no-ops on platforms where they're not supported. e.g. if you add the x11 or wayland features on wasm, they do nothing. They do this by addign a `build.rs`: https://github.com/rust-windowing/winit/blob/fbea75d31f4025825e9bf88b99d7c92e4eb79273/build.rs#L47

We could probably do something similar.

Comment (HyperAlch) [reply Cargo.toml:20]: Sure, sounds good to me

Comment (HyperAlch) [reply Cargo.toml:20]: I forgot to ask, do you want me to create the build.rs file or are you doing it?

Comment (johanhelsing) [reply Cargo.toml:20]: I don't have time to do it today at least. Feel free to give it a shot 👍

Comment (HyperAlch) [reply Cargo.toml:20]: So, I've never worked with WASM before and I can't even get the tests to run on the unaltered main branch of your code.

`cargo test --target wasm32-unknown-unknown` doesn't work. I'm not sure if I should be writing a build.rs to make WASM work if I can't even figure out how to run a WASM test.

What am I doing wrong?

Comment (johanhelsing) [reply Cargo.toml:20]: Running tests for wasm is often a bit tricky. For this project i think i just settled for checking that the examples work. What i do is install wasm-server-runner and configure it to be the runner for wasm globally.

Then you can just do

    cargo run --example basic --target wasm32-unknown-unknown

And it should work

Comment (HyperAlch) [reply Cargo.toml:20]: image

I did what you said with the current main branch and I'm getting this error when it runs

Comment (johanhelsing) [reply Cargo.toml:20]: Thanks, it seems I was a bit quick when I did the bevy 0.10 port. Will be fixed in #22

Comment (HyperAlch) [reply Cargo.toml]: I created a build.rs script like you asked. I think everything works, but please feel free to clone my fork and double check.

Comment (johanhelsing) [reply Cargo.toml]: Looks good to me :) Thanks!

File: src/rocksdb_store.rs (hunk 4 above)
Comment (johanhelsing) [reply src/rocksdb_store.rs]: I went ahead and removed it

**Review by johanhelsing — APPROVED** [review]
Looks good to me now 👍 

Thanks a lot!

**Comment by HyperAlch** [comment]
I've never done a pull request before, does this merge?
image

**Comment by johanhelsing** [comment]
> I've never done a pull request before, does this merge? image

Nope, I'll merge it now. Was just waiting for ci checks to pass.

**Pull request merged**